main/src/ca/uwaterloo/flix/language/phase/optimizer/Inliner.scala

```diff
@@ -515,6 +514,37 @@ object Inliner {
     case _ => false
   }
 
+  /**
+    * Returns an inlined and beta-reduced version of `exp0.sym` if the [[shouldInlineDef]] predicate holds.
+    * Otherwise, returns an [[Expr.ApplyDef]] expression , where the subexpressions have been visited.
+    */
+  private def callSiteInlineDef(exp0: Expr.ApplyDef, ctx0: LocalContext)(implicit sym0: Symbol.DefnSym, sctx: SharedContext, root: OccurrenceAst.Root, flix: Flix): Expr = exp0 match {
+    case Expr.ApplyDef(sym, exps, _, _, _, loc) if shouldInlineDef(root.defs(sym), ctx0) =>
+      val es = exps.map(visitExp(_, ctx0))
+      val defn = root.defs(sym)
+      val ctx = ctx0.copy(subst = Map.empty, currentlyInlining = true)
+      betaReduce(defn.exp, defn.fparams.zip(es), loc, ctx)
+
+    case Expr.ApplyDef(sym, exps, itpe, tpe, eff, loc) =>
+      val es = exps.map(visitExp(_, ctx0))
+      Expr.ApplyDef(sym, es, itpe, tpe, eff, loc)
+  }
+
+  /** Returns `true` if `defn` is not recursive and is either a higher-order function or is a direct call to another function. */
+  private def shouldInlineDef(defn: OccurrenceAst.Def, ctx0: LocalContext)(implicit sym0: Symbol.DefnSym): Boolean = {
+    !ctx0.currentlyInlining && !isRecursive(defn, sym0) && isDirectCall(defn)
+  }
+
+  /** Returns `true` if `defn.sym` is equal to `sym0` */
```

Comment:
The comment does not match the implementation.

Also I dont understand the check.

Reply:
True, this is an error